py/miniconf-mqtt/miniconf/miniconf.py

```diff
@@ -43,103 +41,99 @@ def __init__(self, client, prefix):
-
-    def _on_subscribe(self, _client, mid, _qos, _props):
-        LOGGER.info("Handling subscription for %s", mid)
-        if mid not in self._pending_subscriptions:
-            LOGGER.warning("MID: %s, unexpected subscription", mid)
-            return
-
-        self._pending_subscriptions[mid].set_result(True)
-        del self._pending_subscriptions[mid]
-
-    def _on_message(self, _client, topic, payload, _qos, properties):
-        """Callback function for when messages are received over MQTT.
-
-        Args:
-            _client: The MQTT client.
-            topic: The topic that the message was received on.
-            payload: The payload of the message.
-            _qos: The quality-of-service level of the received packet
-            properties: A dictionary of properties associated with the message.
-        """
-        # Extract request_id corrleation data from the properties
-        try:
-            request_id = properties["correlation_data"][0]
-        except KeyError:
-            LOGGER.info("Discarding message without CD")
-            return
-
-        try:
-            ret, fut = self._inflight[request_id]
-        except KeyError:
-            LOGGER.info("Discarding message with unexpected CD: %s", request_id)
-            return
-
-        # When receiving data not on the specific response topic, the data is some partial result of
-        # another response. Append it to the data collected for the request.
-        if topic != self.response_topic:
-            # Payloads for path values are JSON formatted.
-            response = json.loads(payload)
+        self._listening_task = asyncio.create_task(self.process_messages())
 
-            # Handle get subscription data.
-            ret.append(response)
-            return
-
-        # Payloads for generic responses are UTF8
-        response = payload.decode("utf-8")
 
+    async def process_messages(self):
         try:
-            code = dict(properties["user_property"])["code"]
-        except KeyError:
-            LOGGER.warning("Discarding message without response code user property")
-            return
-
-        # Otherwise, a request has completed with a result. Check the result code and handle it
-        # appropriately.
-        if code == "Continue":
-            ret.append(response)
-            return
-
-        if code == "Ok":
-            if response:
-                ret.append(response)
-            fut.set_result(ret)
-        else:
-            fut.set_exception(
-                MiniconfException(f"Received code: {code}, Message: {response}")
-            )
-
-        del self._inflight[request_id]
-
-    async def _do(self, *args, **kwargs):
+            async with Client(self.broker, protocol=MQTTv5, logger=LOGGER) as client:
+                await client.subscribe(self.response_topic)
+                LOGGER.info("MQTT listener started")
+                async for message in client.messages:
```

Comment:
For some odd reason, this cannot use `self.client`. If you try, the `async for` loop immediately raises a `StopAsyncIteration` exception and the function exits. If you put it in a `while True` loop, it will never yield any messages even though it is properly subscribed.

I suspect there's something wrong about how the client is shared across the async tasks. The workaround here is to use a duplicate MQTT client that only lives in this task.

Reply:
my first guess was windows but you have the selector event loop right.
so I don't know.

Reply:
Its a bug in aiomqtt in how they create the messages generator. If you use the `-d` flag, it causes iteration to happen in the discovery loop, but then fail later. If you don't use the `-d` flag, discovery doesn't happen (so we don't iterate messages), and iteration works later.

I think the problem is that `aiomqtt` is constructing the async iterator during client initialization as opposed to whenever `messages` is requested, so its been exhausted when we call it a second time.

Reply:
PR for the fix is https://github.com/sbtinstruments/aiomqtt/pull/312